custom_metrics/css-variables.js

```diff
@@ -0,0 +1,261 @@
+
+	if (ret) {
+		// Drop duplicates
+		return [...new Set(ret)];
+	}
+}
+
+let visited = new Set();
+
+// Recursively walk a CSSStyleRule or CSSStyleDeclaration
+function walkRule(rule, ret) {
+	if (!rule || visited.has(rule)) {
+		return;
+	}
+
+	visited.add(rule);
+
+	let style, selector;
+
+	if (rule instanceof CSSStyleRule && rule.style) {
+		style = rule.style;
+		selector = rule.selectorText;
+	}
+	else if (rule instanceof CSSStyleDeclaration) {
+		style = rule;
+		selector = "";
+	}
+
+	if (style) {
+		let condition;
+		// mirror properties to add. We add them afterwards, so we don't pointlessly traverse them
+		let additions = {};
+
+		for (let property of style) {
+			let value = style.getPropertyValue(property);
+
+			let containsRef = value.indexOf("var(--") > -1;
+			let setsVar = property.indexOf("--") === 0 && property.indexOf("--" + PREFIX) === -1;
+
+			if (containsRef || setsVar) {
+				if (!condition && rule.parentRule) {
+					condition = [];
+					let r = rule;
+
+					while (r.parentRule?.conditionText) {
+						r = r.parentRule;
+						condition.push({
+							type: r instanceof CSSMediaRule? "media" : "supports",
+							test: r.conditionText
+						});
+					}
+				}
+
+				if (containsRef) {
+					// Set mirror property so we can find it in the computed style
+					additions["--" + PREFIX + property] = value.replace(/var\(--/g, PREFIX + "$&");
+
+					let properties = extractValueProperties(value);
+
+					for (let prop of properties) {
+						let info = ret[prop] = ret[prop] || {get: [], set: []};
+						info.get.push({ usedIn: property, value, selector, condition });
+					}
+				}
+
+				if (setsVar) {
+					let info = ret[property] = ret[property] || {get: [], set: []};
+					info.set.push({ value, selector, condition });
+				}
+
+				// Add class so we can find these later
+				if (selector) {
+					for (let el of document.querySelectorAll(selector)) {
+						el.classList.add(`${PREFIX}element`);
+					}
+				}
+			}
+		}
+
+		// Now that we're done, add the mirror properties
+		for (let property in additions) {
+			style.setProperty(property, additions[property]);
+		}
+	}
+
+	if (rule instanceof CSSMediaRule || rule instanceof CSSSupportsRule) {
+		// rules with child rules, e.g. @media, @supports
+		for (let r of rule.cssRules) {
+			walkRule(r, ret);
+		}
+	}
+}
+
+// Return a subset of the DOM tree that contains variable reads or writes
+function buildGraph() {
+	// Elements that contain variable reads or writes.
+	let elements = new Set(document.querySelectorAll(selector));
+	let map = new Map(); // keep pointers to object for each element
+	let ret = [];
+
+	for (let element of elements) {
+		map.set(element, {element, children: []});
+	}
+
+	for (let element of elements) {
+		let ancestor = element.parentNode.closest?.(selector);
+		let obj = map.get(element);
+
+		if (ancestor) {
+			let o = map.get(ancestor);
+			o.children.push(obj)
```

Comment:
> Do you have any ideas on how I could compress/eliminate them?

First thing that comes to mind is serialize the `obj`, save it in a Set, and test to see if each new object already exists before appending to `o.children`.

Or check the `children` array itself:

```js
if (!o.children.find(child => JSON.stringify(child) == JSON.stringify(obj)))
```

The object properties may not be serialized in the same order, so this may not be a very robust option.

Reply:
@LeaVerou I'll merge this as-is to make sure that we get it into the sync tomorrow. If you're able to optimize this before the end of the month, feel free to file a PR and we'll try to fit it in.

Reply:
Thanks for merging!

> First thing that comes to mind is serialize the obj, save it in a Set, and test to see if each new object already exists before appending to o.children.

The structure is built recursively, so at the time of that push, `obj` is not necessarily in its final state. But I could do something along those lines after the tree is built.